### PR TITLE
Return false if LDAP server is not available

### LINKED ISSUE #30422: LDAP user backend may throw exception if server is not available when counting users
According to the documentation `countUsers()` [returns either an int or false](https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/public/User/Backend/ICountUsersBackend.php#L36), but it does not throw any exception.

However, the LDAP user backend may throw an exception if the server is not available ([`User_Proxy`](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_Proxy.php#L348) calls [`User_LDAP`](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_LDAP.php#L593), which in turn may call [`Access`](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/Access.php#L1021)); it should return `false` instead to match the expected behaviour, as clients of the user backend might not catch exceptions when calling `countUsers()`, which could lead to internal server errors.

### PR DESCRIPTION
**Pull request by amirkx**
Hi
There were two things i could have done.
-  change **Access.php** to not throw exception 
- change the first function above **Access.php** to catch the exception and return `false`

I though it might be cleaner and more consistent to catch the exception in the above function.

This code is submitted for the  following [issue](https://github.com/nextcloud/server/issues/30422).

### CODE REVIEW

**Review by danxuliu — COMMENTED** [review]
@amirkx and @acsfer Thanks a lot for your contribution!

Unfortunately I am not the right person to review it, as I am not familiar with the LDAP code. Therefore, I do not know if other exceptions besides `ServerNotAvailableException` should be caught (maybe other LDAP plugins throw different exceptions? I do not know) or if it would be better to caught the exception somewhere else.

One thing, though; if the exception is caught in `User_Proxy` I guess that it should wrap only the call to `$backend->countUsers();`. Right now it wraps the whole `foreach` block, so the method will return `false` if any backend fails. However, in the existing code if any backend returned `false` the count returned by the other backends was added to the final count, so if any backend fails it might be better to just ignore it and continue with the other backends as well.

In any case, all this should be reviewed by someone else familiar with the LDAP code :-)

**Comment by amirkx** [comment]
Thank you for your feedback. 
I looked at the  `Access.php` file and this is the `countUser()` [function](https://github.com/nextcloud/server/blob/master/apps/user_ldap/lib/Access.php#L1015-1030). 
As you can see the only exception being throw is  `ServerNotAvailableException`.

```php

/**
	 * @param string $filter
	 * @param string|string[] $attr
	 * @param int $limit
	 * @param int $offset
	 * @return false|int
	 * @throws ServerNotAvailableException
	 */
	public function countUsers($filter, $attr = ['dn'], $limit = null, $offset = null) {
		$result = false;
		foreach ($this->connection->ldapBaseUsers as $base) {
			$count = $this->count($filter, [$base], $attr, $limit, $offset);
			$result = is_int($count) ? (int)$result + $count : $result;
		}
		return $result;
	}

```
If this is not a good implementation, please give a hint to how this should be done. Thank you

**Comment by amirkx** [comment]
the tests failed because there was a  `{` mismatch.

**Comment by danxuliu** [comment]
> As you can see the only exception being throw is `ServerNotAvailableException`.

Of course, but that is in `Access.php`. Maybe [any of the LDAP plugins throw other exceptions](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_LDAP.php#L585), but I have no idea. That is why I am not the right person to review this :-)

> And also could you mention the right person who can review the code. Thank you

Sure, I have pinged them.

File: apps/user_ldap/lib/User_Proxy.php [hunk 1]
```diff
@@ -345,10 +346,14 @@ public function hasUserListings() {
 	public function countUsers() {
 		$users = false;
 		foreach ($this->backends as $backend) {
-			$backendUsers = $backend->countUsers();
+			try{
+				$backendUsers = $backend->countUsers();
+			} catch (ServerNotAvailableException $e) {
+					return false;
```
Comment (come-nc) [inline apps/user_ldap/lib/User_Proxy.php:352]: I think you should not return false but ignore this backend, as is done when it returns false.

File: apps/user_ldap/lib/User_Proxy.php [hunk 2]
```diff
@@ -345,10 +346,14 @@ public function hasUserListings() {
...
+					}
 			if ($backendUsers !== false) {
 				$users += $backendUsers;
-			}
+				}
```
Comment (come-nc) [inline apps/user_ldap/lib/User_Proxy.php:356]: Please fix identation

**Review by come-nc — CHANGES_REQUESTED** [review]

**Comment by come-nc** [comment]
Thanks for your contribution, please sign your work when committing!
https://github.com/nextcloud/appstore/blob/master/CONTRIBUTING.md#sign-your-work

**Comment by blizzz** [comment]
@amirkx what is the reasons for this change? I refrain to change the behaviour for we consider the connected LDAP servers a critical component. This change would be inconsistent with the regular behaviour.

**Comment by amirkx** [comment]
I don't know if i have misunderstood the  issue or not. To clarify what i did i'll say what i think what happens and if it's not correct, please correct me.

## problem
If the server is not available (`User_Proxy` calls `User_LDAP`, which in turn may call Access). [refrence in issue description](https://github.com/nextcloud/server/issues/30422)
since we don't catch the exception in [User_proxy](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_Proxy.php#L348) or in [User_LDAP](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_LDAP.php#L593), this may lead to interval server errors.(). [refrence in issue description](https://github.com/nextcloud/server/issues/30422)
## solution
catch the thrown exceptions  in  `User_proxy` and returning `false` instead of an exception.

**Comment by amirkx** [comment]
> This change would be inconsistent with the regular behaviour.

I changed my commit to be more consistent with the existing code.
I also found a familiar pattern in [here](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_Proxy.php#L183).
If i'm misunderstanding please let me know.
@danxuliu

**Comment by blizzz** [comment]
> I don't know if i have misunderstood the issue or not. To clarify what i did i'll say what i think what happens and if it's not correct, please correct me.
> ## problem
> 
> If the server is not available (`User_Proxy` calls `User_LDAP`, which in turn may call Access). [refrence in issue description](https://github.com/nextcloud/server/issues/30422) since we don't catch the exception in [User_proxy](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_Proxy.php#L348) or in [User_LDAP](https://github.com/nextcloud/server/blob/f3dcbfe146782d2c7dec8760651e79605ddc96e7/apps/user_ldap/lib/User_LDAP.php#L593), this may lead to interval server errors.(). [refrence in issue description](https://github.com/nextcloud/server/issues/30422)
> ## solution
> 
> catch the thrown exceptions in `User_proxy` and returning `false` instead of an exception.

Sure, now this is indeed a 5xx error, so it is correct for Nextcloud to act like this.

**Comment by amirkx** [comment]
> Sure, now this is indeed a 5xx error, so it is correct for Nextcloud to act like this.

I'm a little bit confused.
Isn't the whole purpose of this [issue](https://github.com/nextcloud/server/issues/30422) to stop an 5xx error from apprearing to the end user?
@blizzz

**Comment by blizzz** [comment]
> > Sure, now this is indeed a 5xx error, so it is correct for Nextcloud to act like this.
> 
> I'm a little bit confused. Isn't the whole purpose of this [issue](https://github.com/nextcloud/server/issues/30422) to stop an 5xx error from apprearing to the end user? @blizzz

I am equally confused about  the issue :) What i said here, applies there, too. @danxuliu?

**Comment by danxuliu** [comment]
> > > Sure, now this is indeed a 5xx error, so it is correct for Nextcloud to act like this.
> > 
> > I'm a little bit confused. Isn't the whole purpose of this [issue](https://github.com/nextcloud/server/issues/30422) to stop an 5xx error from apprearing to the end user? @blizzz
> 
> I am equally confused about the issue :) What i said here, applies there, too. @danxuliu?

If an app calls `countUsers()` but it does not catch the exceptions thrown by it it could happen that after misconfiguring the LDAP server trying to access the settings always ends in a 5XX error, so the configuration can not be fixed using the WebUI.

This happens with the support app. I would have caught the exception there, but then I noticed that `IUserManager` documentation did not mention anything about `countUsers()` throwing exceptions. Therefore I assumed that, if there was a problem counting the users, something would be written to the logs, but no exception would be thrown by `countUsers()` :-) However, when I found this @blizzz was on vacation; as I could not ask you I just opened the issue to not forget about it ;-)

If it is correct for Nextcloud to end in a 5XX error when counting users is not possible and the error is not explicitly handled then there is nothing to fix in the LDAP plugin.

A hint in the public API documentation would be nice, though, although it may be there already and I am not aware of it (for example, in the developer manual or something like that because it is a behaviour common to any public Nextcloud API).

**Comment by blizzz** [comment]
> > > > Sure, now this is indeed a 5xx error, so it is correct for Nextcloud to act like this.
> > > 
> > > 
> > > I'm a little bit confused. Isn't the whole purpose of this [issue](https://github.com/nextcloud/server/issues/30422) to stop an 5xx error from apprearing to the end user? @blizzz
> > 
> > 
> > I am equally confused about the issue :) What i said here, applies there, too. @danxuliu?
> 
> If an app calls `countUsers()` but it does not catch the exceptions thrown by it it could happen that after misconfiguring the LDAP server trying to access the settings always ends in a 5XX error, so the configuration can not be fixed using the WebUI.
> This happens with the support app. I would have caught the exception there, but then I noticed that `IUserManager` documentation did not mention anything about `countUsers()` throwing exceptions. Therefore I assumed that, if there was a problem counting the users, something would be written to the logs, but no exception would be thrown by `countUsers()` :-) However, when I found this @blizzz was on vacation; as I could not ask you I just opened the issue to not forget about it ;-)

It is very unfortunate that the support app counts the users on a user facing action in general as it may be expensive with external backends. It would be better if the support app does it in the background, and maybe once or a few times a day is enough? Or it may resort to `countSeenUsers`.

> If it is correct for Nextcloud to end in a 5XX error when counting users is not possible and the error is not explicitly handled then there is nothing to fix in the LDAP plugin.

Now counting users is not that sensitive, probably, but there are certainly issues that may pop up as we painfully experiences in the past, therefore LDAP generally shuts down the server – like missing DB or Redis does.

> A hint in the public API documentation would be nice, though, although it may be there already and I am not aware of it (for example, in the developer manual or something like that because it is a behaviour common to any public Nextcloud API).

Certainly there needs to be better documentation indeed. What's given there is too shallow indeed.

**Comment by amirkx** [comment]
Good to hear there is no problem to solve :)
should i close the PR?

If possible I would like to  help with the Documentation that you mentioned. If you could kindly point me to the documentation file. @blizzz

**Pull request closed without merging**